File: be/app/routers/alignments.py
"""Alignments router - create, list, align, resolve, conflicts"""

import logging
from threading import Thread

from fastapi import APIRouter, Depends, HTTPException, Query, UploadFile, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.dependencies import get_current_user
from app.models.alignment import AlignmentState
from app.models.user import User
from app.schemas.alignment import (
    AlignmentCreate,
    AlignmentOut,
    AlignNext,
    AlignStart,
    ResolveRequest,
)
from app.services import alignment_service, processing_service
from app.services.document_service import get_document_by_guid

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/aligner/alignments", tags=["alignments"])


@router.get("/", response_model=list[AlignmentOut])
def list_alignments(
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    return alignment_service.list_alignments(db, user.id)


@router.post("/", response_model=AlignmentOut, status_code=status.HTTP_201_CREATED)
def create_alignment(
    data: AlignmentCreate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    doc_from = get_document_by_guid(db, user.id, data.document_from_guid)
    doc_to = get_document_by_guid(db, user.id, data.document_to_guid)
    if not doc_from or not doc_to:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Source or target document not found",
        )

    alignment = alignment_service.create_alignment(
        db, user.id, doc_from, doc_to, data.name
    )
    return alignment


@router.delete("/{guid}", status_code=status.HTTP_204_NO_CONTENT)
def delete_alignment(
    guid: str,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    alignment = alignment_service.get_alignment(db, user.id, guid)
    if not alignment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Alignment not found")
    alignment_service.delete_alignment(db, user.id, guid)


@router.post("/{guid}/align")
def start_alignment(
    guid: str,
    data: AlignStart,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    alignment = alignment_service.get_alignment(db, user.id, guid)
    if not alignment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Alignment not found")

    if data.align_all:
        alignment_service.update_state(
            db, alignment.id, AlignmentState.INIT, 0, alignment.total_batches
        )
    alignment_service.update_state(db, alignment.id, AlignmentState.IN_PROGRESS)

    # Run in background thread to not block the response
    thread = Thread(
        target=processing_service.start_alignment,
        args=(user.id, alignment, data),
        daemon=True,
    )
    thread.start()

    return {"status": "started"}


@router.post("/{guid}/align/next")
def align_next(
    guid: str,
    data: AlignNext,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    alignment = alignment_service.get_alignment(db, user.id, guid)
    if not alignment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Alignment not found")

    alignment_service.update_state(db, alignment.id, AlignmentState.IN_PROGRESS)

    thread = Thread(
        target=processing_service.align_next,
        args=(user.id, alignment, data),
        daemon=True,
    )
    thread.start()

    return {"status": "started"}


@router.post("/{guid}/align/stop")
def stop_alignment(
    guid: str,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    alignment = alignment_service.get_alignment(db, user.id, guid)
    if not alignment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Alignment not found")

    processing_service.stop_alignment(db, alignment)
    return {"status": "stopped"}


@router.post("/{guid}/resolve")
def resolve_conflicts(
    guid: str,
    data: ResolveRequest,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    alignment = alignment_service.get_alignment(db, user.id, guid)
    if not alignment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Alignment not found")

    alignment_service.update_state(db, alignment.id, AlignmentState.IN_PROGRESS)

    thread = Thread(
        target=processing_service.resolve_conflicts,
        args=(user.id, alignment, data),
        daemon=True,
    )
    thread.start()

    return {"status": "started"}


@router.get("/{guid}/conflicts")
def get_conflicts(
    guid: str,
    handle_edges: str = Query("none"),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    alignment = alignment_service.get_alignment(db, user.id, guid)
    if not alignment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Alignment not found")

    return processing_service.get_conflicts(user.id, alignment, handle_edges)


@router.get("/{guid}/conflicts/{conflict_id}")
def show_conflict(
    guid: str,
    conflict_id: int,
    handle_edges: str = Query("none"),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    alignment = alignment_service.get_alignment(db, user.id, guid)
    if not alignment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Alignment not found")

    return processing_service.show_conflict(user.id, alignment, conflict_id, handle_edges)


@router.post("/{guid}/visualize")
def update_visualization(
    guid: str,
    batch_ids: list[int] = [],
    update_all: bool = False,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    alignment = alignment_service.get_alignment(db, user.id, guid)
    if not alignment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Alignment not found")

    processing_service.update_visualization(user.id, alignment, batch_ids, update_all)
    return {"status": "ok"}


@router.post("/{guid}/proxy/{direction}", response_model=AlignmentOut)
async def upload_proxy(
    guid: str,
    direction: str,
    file: UploadFile,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    if direction not in ("from", "to"):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="direction must be 'from' or 'to'",
        )
    alignment = alignment_service.get_alignment(db, user.id, guid)
    if not alignment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Alignment not found")

    content = (await file.read()).decode("utf-8")
    alignment_service.upload_proxy(user.id, alignment, direction, content)
    return alignment_service.update_proxy_loaded(db, alignment.id, direction)


@router.get("/{guid}/progress", response_model=AlignmentOut)
def get_progress(
    guid: str,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    alignment = alignment_service.get_alignment(db, user.id, guid)
    if not alignment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Alignment not found")

    db.refresh(alignment)
    return alignment

File: fe/src/api/alignments.ts
import { apiFetch } from './client'

export const AlignmentState = {
  INIT: 0,
  IN_PROGRESS: 1,
  IN_PROGRESS_DONE: 2,
  DONE: 3,
  ERROR: 4,
} as const

export type AlignmentStateValue = (typeof AlignmentState)[keyof typeof AlignmentState]

export interface AlignmentOut {
  id: number
  guid: string
  name: string
  lang_from: string
  lang_to: string
  state: AlignmentStateValue
  curr_batches: number
  total_batches: number
  is_deleted: boolean
  is_uploaded: boolean
  proxy_from_loaded: boolean
  proxy_to_loaded: boolean
  document_from_id: number
  document_to_id: number
  created_at: string
}

export interface AlignmentCreate {
  name: string
  document_from_guid: string
  document_to_guid: string
}

export interface AlignStartParams {
  align_all?: boolean
  batch_ids?: number[]
  batch_shift?: number
  window?: number
  use_proxy_from?: boolean
  use_proxy_to?: boolean
}

export interface AlignNextParams {
  amount?: number
  batch_shift?: number
  window?: number
  use_proxy_from?: boolean
  use_proxy_to?: boolean
}

export interface ResolveConflictsParams {
  batch_ids?: number[]
  use_proxy_from?: boolean
  use_proxy_to?: boolean
  handle_start?: boolean
  handle_finish?: boolean
}

export interface VisualizationParams {
  batch_ids?: number[]
  update_all?: boolean
}

export interface Conflict {
  id: number
  batch_id: number
  type: string
  data: Record<string, unknown>
}

export interface ConflictsResponse {
  conflicts: Conflict[]
}

export interface ConflictDetail {
  id: number
  batch_id: number
  type: string
  left: string[]
  right: string[]
  candidates: string[][]
  data: Record<string, unknown>
}

export function listAlignments() {
  return apiFetch<AlignmentOut[]>('/api/aligner/alignments/')
}

export function createAlignment(data: AlignmentCreate) {
  return apiFetch<AlignmentOut>('/api/aligner/alignments/', {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export function deleteAlignment(guid: string) {
  return apiFetch<void>(`/api/aligner/alignments/${guid}`, { method: 'DELETE' })
}

export function startAlignment(guid: string, data: AlignStartParams) {
  return apiFetch<void>(`/api/aligner/alignments/${guid}/align`, {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export function alignNext(guid: string, data: AlignNextParams) {
  return apiFetch<void>(`/api/aligner/alignments/${guid}/align/next`, {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export function stopAlignment(guid: string) {
  return apiFetch<void>(`/api/aligner/alignments/${guid}/align/stop`, { method: 'POST' })
}

export function resolveConflicts(guid: string, data: ResolveConflictsParams) {
  return apiFetch<void>(`/api/aligner/alignments/${guid}/resolve`, {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export function getConflicts(guid: string) {
  return apiFetch<ConflictsResponse>(`/api/aligner/alignments/${guid}/conflicts`)
}

export function getConflictDetail(guid: string, id: number) {
  return apiFetch<ConflictDetail>(`/api/aligner/alignments/${guid}/conflicts/${id}`)
}

export function updateVisualization(guid: string, data: VisualizationParams) {
  return apiFetch<void>(`/api/aligner/alignments/${guid}/visualize`, {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export function getProgress(guid: string) {
  return apiFetch<AlignmentOut>(`/api/aligner/alignments/${guid}/progress`)
}
